### PR TITLE
Llama 3 405B on Trillium instructions (experimental docker)

### PR DESCRIPTION
**Pull request by tengyifei**

### CODE REVIEW

**Review by qihqi — APPROVED** [review]

File: Training/TPU-Trillium/Llama3-405B-PyTorch/train.sh [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  --tokenizer_name meta-llama/Meta-Llama-3.1-405B \
+  --block_size "$SEQ_LENGTH" \
+  --optim adafactor \
+  --save_strategy no \
+  --logging_strategy no \
+  --torch_dtype bfloat16 \
+  --dataloader_drop_last yes \
+  --flash_attention \
+  --spmd_2d_sharding 4 \
+  --max_steps "$MAX_STEPS" \
```
Comment (qihqi) [inline Training/TPU-Trillium/Llama3-405B-PyTorch/train.sh:35]: there seems to be an extra \

File: Training/TPU-Trillium/Llama3-405B-PyTorch/host.sh [hunk 2]
```diff
@@ -0,0 +1,32 @@
+#!/bin/bash
+
+DOCKER_IMAGE=us-central1-docker.pkg.dev/deeplearning-images/reproducibility/pytorch-xla/llama3-405b:nightly-sep28
+
+worker_id=$(curl -s "http://metadata.google.internal/computeMetadata/v1/instance/attributes/agent-worker-number" -H 'Metadata-Flavor: Google')
+
+cat >> /dev/null <<EOF
```
Comment (qihqi) [inline Training/TPU-Trillium/Llama3-405B-PyTorch/host.sh:7]: what is this doing?

File: Training/TPU-Trillium/Llama3-405B-PyTorch/README.md [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+```bash
+git clone --depth 1 https://github.com/gclouduniverse/reproducibility.git
+cd reproducibility/Training/TPU-Trillium/Llama3-405B-PyTorch
+```
+
+2. Edit `env.sh` to add the hugging face token and/or setup the training parameters.
+
+```bash
+# add your hugging face token
+HF_TOKEN=hf_***
```
Comment (qihqi) [inline Training/TPU-Trillium/Llama3-405B-PyTorch/README.md:48]: add a line here to ask the user to also set the HF_TOKEN in env.sh

File: Training/TPU-Trillium/Llama3-405B-PyTorch/host.sh (hunk 2 above)
Comment (tengyifei) [reply Training/TPU-Trillium/Llama3-405B-PyTorch/host.sh:7]: I don't really know -- I copied it from existing scripts. I know at a high level this runs the training command in a docker container.

File: Training/TPU-Trillium/Llama3-405B-PyTorch/train.sh [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+mkdir -p "${LOCAL_DIR}/output"
+mkdir -p "${LOCAL_DIR}/plugins"
+mkdir -p "${LOCAL_DIR}/cache"
+
+unset LD_PRELOAD
+
+
+cd transformers/
+
+
```
Comment (miladm) [inline Training/TPU-Trillium/Llama3-405B-PyTorch/train.sh:16]: Let's explicitly call out `minibatch=False`

File: Training/TPU-Trillium/Llama3-405B-PyTorch/README.md [hunk 5]
```diff
@@ -0,0 +1,77 @@
+# Instructions for training Llama 3 405B on Trillium TPU
+
+This user guide provides a concise overview of the essential steps required to
+run Hugging Face (HF) Llama 3 405B training on Trillium TPUs.
```
Comment (miladm) [inline Training/TPU-Trillium/Llama3-405B-PyTorch/README.md:4]: Suggesting the following update:

Note: the current docker supports Single Pod v6e. The multipod solution will be available in an upcoming update soon.

File: Training/TPU-Trillium/Llama3-405B-PyTorch/env.sh [hunk 6]
```diff
@@ -0,0 +1,19 @@
+# Uncomment below to set the Huggingface token
+HF_TOKEN=hf_***
+PJRT_DEVICE=TPU
+XLA_IR_DEBUG=1
+XLA_HLO_DEBUG=1
+PROFILE_EPOCH=0
+PROFILE_STEP=10
+PROFILE_DURATION_MS=240000
+PROFILE_LOGDIR=/tmp/home/profile
```
Comment (miladm) [inline Training/TPU-Trillium/Llama3-405B-PyTorch/env.sh:9]: are tensorboard artifacts + tpu-info installed in this docker? @tengyifei

**Review by miladm — APPROVED** [review]
Thanks.
Added a few comments for you to address
Plz confirm >1 person has had success running this image

Approving to unblock

File: Training/TPU-Trillium/Llama3-405B-PyTorch/env.sh (hunk 6 above)
Comment (tengyifei) [reply Training/TPU-Trillium/Llama3-405B-PyTorch/env.sh:9]: tpu-info is installed. However, this training command doesn't produce any tensorboard artifacts. Even if someone installed tensorboard, they won't get any outputs. I'll look into what's needed to get those.

**Review by miladm — APPROVED** [review]

File: Training/TPU-Trillium/Llama3-405B-PyTorch/env.sh (hunk 6 above)
Comment (miladm) [reply Training/TPU-Trillium/Llama3-405B-PyTorch/env.sh:9]: are you saying profiling is not operational in this docker? if so, we can postpone tensorboard installation to the next docker and move fwd.

**Comment by miladm** [comment]
Feel free to clarify in README that this docker is in experimental stage @tengyifei

**Review by qihqi — APPROVED** [review]

**Review by qihqi — APPROVED** [review]

**Comment by qihqi** [comment]
I also ran this instructions and it works.

**Pull request merged**